python/lsst/daf/butler/core/s3utils.py

```diff
@@ -0,0 +1,178 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("s3CheckFileExists", "bucketExists")
+
+try:
+    import boto3
+except ImportError:
+    boto3 = None
+
+from lsst.daf.butler.core.location import ButlerURI, Location
+
+
+def s3CheckFileExistsGET(client, bucket, filepath):
+    """Returns (True, filesize) if file exists in the bucket and (False, -1) if
+    the file is not found.
+
+    Parameters
+    ----------
+    client : `boto3.client`
+        S3 Client object to query.
+    bucket : `str`
+        Name of the bucket in which to look.
+    filepath : `str`
+        Path to file.
+
+    Returns
+    -------
+    (`bool`, `int`) : `tuple`
+       Tuple (exists, size). If file exists (True, filesize)
+       and (False, -1) when the file is not found.
+
+    Notes
+    -----
+    A Bucket GET request will be charged against your account. This is on
+    average 10x cheaper than a LIST request (see `s3CheckFileExistsLIST`
+    function) but can be up to 90% slower whenever additional work is done with
+    the s3 client. See PR:
+        ``https://github.com/boto/botocore/issues/1248``
+    for details, in boto3 versions >=1.11.0 the loss of performance should not
+    be an issue anymore.
+    S3 Paths are sensitive to leading and trailing path separators.
+    """
+    try:
+        obj = client.head_object(Bucket=bucket, Key=filepath)
+        return (True, obj['ContentLength'])
+    except client.exceptions.ClientError as err:
+        if err.response["ResponseMetadata"]["HTTPStatusCode"] == 404:
+            return (False, -1)
+        raise
+
+
+def s3CheckFileExistsLIST(client, bucket, filepath):
+    """Returns (True, filesize) if file exists in the bucket and (False, -1) if
+    the file is not found.
+
+    Parameters
+    ----------
+    client : `boto3.client`
+        S3 Client object to query.
+    bucket : `str`
+        Name of the bucket in which to look.
+    filepath : `str`
+        Path to file.
+
+    Returns
+    -------
+    (`bool`, `int`) : `tuple`
+       Tuple (exists, size). If file exists (True, filesize)
+       and (False, -1) when the file is not found.
+
+    Notes
+    -----
+    You are getting charged for a Bucket LIST request. This is on average 10x
+    more expensive than a GET request (see `s3CheckFileExistsGET` fucntion) but
+    can be up to 90% faster whenever additional work is done with the s3
+    client. For details, see PR:
+        ``https://github.com/boto/botocore/issues/1248``
+    Boto3 versions >=1.11.0 the loss of performance should not be an issue
+    anymore.
+    """
+    response = client.list_objects_v2(
+        Bucket=bucket,
+        Prefix=filepath
+    )
+    # Hopefully multiple identical files will never exist?
+    matches = [x for x in response.get('Contents', []) if x["Key"] == filepath]
```

Comment:
That is exactly correct. Since there are no directories, the name of the key is its full path. Since we are checking for a file and not directory I expect never to hit duplicates. I added required comments into the Notes section. This function is itching to get removed anyhow, as soon as conda package index catches up to newer version of boto3 since the change was recent-ish. Is that ok?

Reply:
So it's on pypi but not yet on conda?